[All] Use IReflectableType when accessing the Registrar

praeclarum opened the pull request:
### Description of Change ###

Everywhere `Registrar.Registered.GetHandler` is called, try to use the `IReflectableType` to get type information instead of `GetType`.

This allows dynamic objects that implement their own types to be used by the Registrar.

There is already some precedent for support of dynamic objects: `BindingExpression.ApplyCore` does this same kind of check. This patch extends that support to the Registrar.

### API Changes ###

None

### Behavioral Changes ###

Users will be able to register dynamic objects in the Registrar where before they could not.

### PR Checklist ###

- [x] Has tests (2 new tests)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

dnfclas commented:
@praeclarum,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

StephaneDelcroix commented on Xamarin.Forms.Platform.Android/Cells/BaseCellView.cs:
```diff
@@ -190,7 +190,10 @@ public void SetRenderHeight(double height)
+			var rt = source as System.Reflection.IReflectableType;
+			var t = rt != null ? rt.GetTypeInfo().AsType() : source.GetType();
+
+			if (source != null && (handler = Registrar.Registered.GetHandler<IImageSourceHandler>(t)) != null)
```
the check for `source != null` happens to late. L194 is potentially accessing a `null` instance

what about 
```
var t = (source as IReflectableType) ?.GetTypeInfo().AsType() ?? source?.GetType();
if (t != null && ....)
```

praeclarum replied:
Yeah I definitely screwed up there. I will be more careful on v2.

StephaneDelcroix commented on Xamarin.Forms.Platform.Android/Extensions/ImageViewExtensions.cs:
```diff
@@ -31,7 +31,10 @@ internal static class ImageViewExtensions
+			var rt = source as System.Reflection.IReflectableType;
+			var t = rt != null ? rt.GetTypeInfo().AsType() : source.GetType();
+
+			if (source != null && (handler = Internals.Registrar.Registered.GetHandler<IImageSourceHandler>(t)) != null)
```
_idem_

praeclarum commented:
@StephaneDelcroix I agree with you and realized after I did all of this.

I would be happy to kill this PR and start over with a new function on the Registrar that hid all this ugliness.

praeclarum commented:
@StephaneDelcroix Do you have any suggestions for the name of `GetHandlerXXX`? Right now there's `GetHandler` and `GetHandlerType`. I propose to add these:

* `GetHandlerForObject`
* `GetHandlerTypeForObject`

If you don't like those, how about:

* `GetObjectHandler`
* `GetObjectHandlerType`

StephaneDelcroix commented:
I prefer the first pair. You @jassmith ??

praeclarum commented:
@StephaneDelcroix I have rewritten the PR to use the new methods. I have also added docs and unit tests for them.

Still happy to rename if @jassmith doesn't like them.

jassmith commented:
GetHandlerForObject is fine, rest of the PR looks fine, need to get the test ran then we can merge.

StephaneDelcroix commented:
@praeclarum here's the build failure

```
Compilation error: Xamarin.Forms.Platform.Android\Xamarin.Forms.Platform.Android.csproj 
Cells\ViewCellRenderer.cs(97, 139): error CS0246: The type or namespace name 'obj' could not be found (are you missing a using directive or an assembly reference?) ```

StephaneDelcroix commented on Xamarin.Forms.Platform.Android/Cells/ViewCellRenderer.cs:
```diff
@@ -92,8 +92,10 @@ public void Update(ViewCell cell)
-				if (renderer != null && renderer.GetType() == viewHandlerType)
+				var viewHandlerType = Registrar.Registered.GetHandlerTypeForObject(cell.View) ?? typeof(Platform.DefaultRenderer);
+				var reflectableType = renderer as System.Reflection.IReflectableType;
+				var rendererType = reflectableType != null ? reflectableType.GetTypeInfo().AsType() : (renderer != null ? renderer.GetType() : typeof(obj));
```
`typeof(obj)` ???

praeclarum replied:
Yep not sure how I missed this. Fixing...

hartez commented:
@praeclarum This doesn't seem to be compiling.

praeclarum commented:
@hartez Sorry for the delay, fixing...

rmarinho commented:
build

rmarinho commented:
@praeclarum did you pushed the fix ?
